cmd/cosign/cli/verify.go

```diff
@@ -59,6 +59,9 @@ against the transparency log.`,
   # verify image with local certificate and certificate chain
   cosign verify --cert cosign.crt --cert-chain chain.crt <IMAGE>
 
+  # verify image with local root certificate
+  cosign verify --root-cert root.crt <IMAGE>
```

Comment:
@znewman01 I suspect you might disagree, but I don’t see a reason to have many different ways to pass the root. 
* via cert-chain
* Via the env var
* TUF root
* This option now

I don’t see the confusion with using cert-chain. It’s documented that you can either pass a root and verify should pick up any intermediates from the image annotation, or you pass the whole chain if intermediates aren’t stored on the image.

Reply:
The other issue with this is its being added only to verify and not the other verify-* commands. If this is only being added to support attach, the flag should be scoped as such, but I think that would cause confusion.

Reply:
I’d propose we either:
* Stick with passing a root via the cert chain flag 
* refactor all verify commands to pass roots and intermediates separately, ala openssl-style (openssl is not what I’d use an example of good UX fwiw)

Reply:
One more idea: We can now link environment variables to flags, so a cleaner approach that would reduce duplication would be to link SIGSTORE_ROOT_FILE to a CLI flag for the trust root. For names, I’d pick “SIGSTORE_CA_TRUST_ROOT” and —ca-trust-root, what you proposed Zack. I also like this because it doesn’t treat fulcio differently than other CAs. 

I can get this out tomorrow if this sounds good.

Reply:
As part of the given PR we want to solve the shortcoming we identified in cosign for the following use case.
****Use case** : A corporate has a private CA which is used for certificate generation and signing. The corporate want to use there existing infrastructure for signature signing and certificate generation but want to use cosign tool to upload the signature, cert, cert chain in registry and also to verify the image content at the verifier side.**

while investigating the usage of cosign in the scenario of private CA we found following short comings
1. The attach signature command was not allowing attachment of certificate and Cert chain in the registry
2. if we use cosign sign command along with cert and cert-chain the cert and cert chain get updated in registry but when we use Verify command we have to explicitly pass  the cert and cert-chain as an argument even if these things are already attached in the registry along with the image. One way to do this is to first download the cert and cert-chain from registry and then use them as the argument of verify command.
3. There should be a source of trust at Verifier side which should verify the cert-chain , cert and signature, as any adversary can modify the data in the registry by updating signature, cert and cert chain so we can not just rely on the data present in the registry the source of trust should be with verifier.

So to fix these short coming we propose following enhancement in this PR.
1. Attach command has been modified to allow attachment of Cert and cert chain(Including root certificate) along with signature and payload
2. Verify command has a new argument "--root-cert" which Verifier can use to verify the Cert-chain, cert and signature. The root certificate which we will used by verifier will be provided to the verifier using out of band mechanisms. and  Verifier will use the trusted root certificate to verify the cert chain. 

This is very similar to the fulcio root based verification approach where cosign fetch the root cert from fulcio to verify the stored cert-chain and cert attached with the signature. the only difference here is that use is using a locally stored root certificate to validate the certificate chain instead of using public CA like fulcio.

Reply:
Yep, I’m just talking about how you pass the root certificate to verify. This issue is a little orthogonal to this PR, we have another issue open to make “cert-chain” clearer, but adding another flag to provide a root will exacerbate the issue. 

To unblock this PR, I would just stick with what options we already have for passing root material and not add an additional flag.

Reply:
Thanks for the update. so if i understood you correctly the proposal is that we would have an environment flag "SIGSTORE_CA_TRUST_ROOT" which will be set to the path of the local trusted root certificate. and than user can just 
export SIGSTORE_CA_TRUST_ROOT=<PATH of trusted root cert>
execute the following command
cosign verify <IMAGE>
This will pick the Trusted root cert from the Env variable to a internal flag "—ca-trust-root" and use that to verify the cert-chain , cert and signature stored in the registry.
so this seems ok with me, as this will solve the requirement we need to be supported in the cosign tool. kindly clarify if my understanding is correct.

Reply:
I would propose that for this PR, you remove the added `root-cert` flags and rely on the existing `certificate-chain` flag to verify contents. You can remove your changes to `verify.go`. This keeps the PR small and focused just on improving attach.

There's more that we need to think through for the user experience for verifying certificates, and I wouldn't want to rush to add new verification flags if it will be confusing for users.

Reply:
i tried passing the root-cert with cert-chain option. but it tries to search for fulico root cert. so the command fails
====
C:\Users\musharma\Desktop\cosignCert>cosign.exe verify --cert-chain C:\Users\musharma\Desktop\cosignCert\rootCA.crt harbor.ncdfp3.bpti.dyn.nesc.nokia.net/cosign-demo/fp3testcnf/orbs/orb-atp@sha256:8cbd5009dff7db858c58d54c051fd99bfe7128aa4345d2cf5480dd55b9e7b711 --allow-insecure-registry --insecure-ignore-sct --insecure-skip-tlog-verify --certificate-identity-regexp .* --certificate-oidc-issuer-regexp .*
**Warning** Skipping tlog verification is an insecure practice that lacks of transparency and auditability verification for the signature.
Error: getting Fulcio roots: error getting targets: error getting target fulcio.crt.pem by usage: open C:\Users\musharma\.sigstore\root\targets\fulcio.crt.pem: The system cannot find the file specified.
main.go:62: error during command execution: getting Fulcio roots: error getting targets: error getting target fulcio.crt.pem by usage: open C:\Users\musharma\.sigstore\root\targets\fulcio.crt.pem: The system cannot find the file specified.
====
The working scenario with existing mechanism provided by cosign is  to use env variable SIGSTORE_ROOT_FILE
to the Trusted local root certificate and use cosign verify command without any options.
===
C:\Users\musharma\Desktop\cosignCert>cosign.exe verify harbor.ncdfp3.bpti.dyn.nesc.nokia.net/cosign-demo/fp3testcnf/orbs/orb-atp@sha256:8cbd5009dff7db858c58d54c051fd99bfe7128aa4345d2cf5480dd55b9e7b711 --allow-insecure-registry --insecure-ignore-sct --insecure-skip-tlog-verify --certificate-identity-regexp .* --certificate-oidc-issuer-regexp .*
**Warning** Skipping tlog verification is an insecure practice that lacks of transparency and auditability verification for the signature.

Verification for harbor.ncdfp3.bpti.dyn.nesc.nokia.net/cosign-demo/fp3testcnf/orbs/orb-atp@sha256:8cbd5009dff7db858c58d54c051fd99bfe7128aa4345d2cf5480dd55b9e7b711 --
The following checks were performed on each of these signatures:
  - The cosign claims were validated
  - Any certificates were verified against the Fulcio roots.

[{"critical":{"identity":{"docker-reference":"harbor.ncdfp3.bpti.dyn.nesc.nokia.net/cosign-demo/fp3testcnf/orbs/orb-atp"},"image":{"docker-manifest-digest":"sha256:8cbd5009dff7db858c58d54c051fd99bfe7128aa4345d2cf5480dd55b9e7b711"},"type":"cosign container image signature"},"optional":{"Subject":"foo@example.com"}}]

C:\Users\musharma\Desktop\cosignCert>echo %SIGSTORE_ROOT_FILE%
C:\Users\musharma\Desktop\cosignCert\rootCA.crt
====
As discussed before I have now reverted the changes done in the following files
cmd/cosign/cli/options/certificate.go
cmd/cosign/cli/verify.go
cmd/cosign/cli/verify/verify.go